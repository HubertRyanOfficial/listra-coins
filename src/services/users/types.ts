export type UserPostType = {
  email: string;
  password: string;
  name: string;
  balance: number;
  created_at: number;
  profileImage: string | null;
};

export type UserGetType = UserPostType & {
  id: string;
};
