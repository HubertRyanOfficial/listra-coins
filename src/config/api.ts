import axios from "axios";
import * as Device from "expo-device";

const api = axios.create({
  baseURL: !Device.isDevice
    ? "http://localhost:3000"
    : "http://192.168.0.104:3000",
});

export default api;
