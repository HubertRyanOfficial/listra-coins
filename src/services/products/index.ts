import api from "@/config/api";
import { Product, ShoppingItemType } from "./types";

export const getAllProducts = async (userId: string) => {
  const responses = await Promise.all([
    api.get<Product[]>("/products"),
    api.get<ShoppingItemType[]>("/shopping"),
  ]);
  const products = responses[0].data;
  const shoppingItems = responses[1].data;

  return products.map((product) => {
    const hasInCart = shoppingItems.find(
      (item) => item.product_id == product.id && item.user_id == userId
    )
      ? true
      : false;

    return {
      ...product,
      in_cart: hasInCart,
    };
  });
};
