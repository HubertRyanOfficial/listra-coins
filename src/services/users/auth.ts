import api from "@/config/api";
import { UserGetType } from "./types";

export const loginUser = async (email: string, password: string) => {
  const response = await api.get<UserGetType[]>("/users");
  const users = response.data;

  const currentUser = users.find((user) => user.email === email);

  if (currentUser) {
    if (currentUser.password === password) {
      return {
        id: currentUser.id,
        email: currentUser.email,
        name: currentUser.name,
        balance: currentUser.balance,
        created_at: currentUser.created_at,
      };
    } else {
      throw "incorrect-password";
    }
  } else {
    throw "user-not-exists";
  }
};
