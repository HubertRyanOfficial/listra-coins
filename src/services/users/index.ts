import api from "@/config/api";
import { UserGetType, UserPostType } from "./types";

export const createUser = async (data: UserPostType) => {
  const response = await api.get<UserGetType[]>("/users");

  const getCurrentUser = response.data.find(
    (user) => user.email === data.email
  );

  if (getCurrentUser) {
    throw "user-already-exists";
  }

  const user = await api.post("/users", data);
  return user.data;
};

export const updateUser = async (data: Partial<UserPostType>, userId) => {
  const user = await api.get(`/users/${userId}`);
  const response = await api.put<UserGetType>(`/users/${userId}`, {
    ...user.data,
    ...data,
  });
  return response.data;
};
