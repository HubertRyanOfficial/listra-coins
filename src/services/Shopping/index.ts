import api from "@/config/api";
import { Product } from "../products/types";

export const addToCart = async (product: Product, userId: string) => {
  const date = new Date();
  const response = await api.post("/shopping", {
    product_id: product.id,
    user_id: userId,
    quantity: 1,
    total_paid: product.price,
    created_at: date.valueOf(),
  });

  return response.data;
};
