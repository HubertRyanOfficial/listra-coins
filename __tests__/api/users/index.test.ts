import { createUser, updateUser } from "@/services/users";
import api from "@/config/api";
import MockAdapter from "axios-mock-adapter";

const mock = new MockAdapter(api);
const date = new Date().valueOf();

const mockUsers = [
  {
    id: "1",
    email: "user1@example.com",
    password: "password1",
    name: "User 1",
    balance: 100,
    created_at: date,
    profileImage: null,
    notificationToken: null,
  },
  {
    id: "2",
    email: "user2@example.com",
    password: "password2",
    name: "User 2",
    balance: 200,
    created_at: date,
    profileImage: null,
    notificationToken: null,
  },
];

describe("createUser", () => {
  test("creates new user correctly", async () => {
    const newUser = {
      email: "newuser@example.com",
      password: "password",
      name: "New User",
      balance: 0,
      created_at: date,
      profileImage: null,
      notificationToken: null,
    };

    mock.onGet("/users").reply(200, mockUsers);
    mock.onPost("/users").reply(200, { id: "3", ...newUser });

    const createdUser = await createUser(newUser);
    expect(createdUser).toEqual({ id: "3", ...newUser });
  });

  test("throws error for existing user", async () => {
    mock.onGet("/users").reply(200, mockUsers);

    let existingUser = mockUsers[0];
    delete existingUser.id;

    await expect(createUser(existingUser)).rejects.toThrowErrorMatchingSnapshot(
      "user-already-exists"
    );
  });
});

describe("updateUser", () => {
  test("updates user correctly", async () => {
    const userId = "1";
    const updatedData = { name: "Updated User" };

    mock.onGet(`/users/${userId}`).reply(200, { data: mockUsers[0] });
    mock
      .onPut(`/users/${userId}`)
      .reply(200, { id: userId, ...mockUsers[0], ...updatedData });

    const updatedUser = await updateUser(updatedData, userId);

    expect(updatedUser).toEqual({
      id: userId,
      ...mockUsers[0],
      ...updatedData,
    });
  });
});
