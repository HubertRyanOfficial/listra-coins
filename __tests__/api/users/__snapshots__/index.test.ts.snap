// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createUser throws error for existing user: user-already-exists 1`] = `undefined`;
